proposals/0293-extend-property-wrappers-to-function-and-closure-parameters.md

```diff
@@ -504,9 +522,77 @@ Keeping the property-wrapper storage type private is consistent with how propert
+
+### Explicit spelling for API-level property wrappers
+
+The scope of what is considered an API-level property wrapper is very limited in this proposal, and the external effect of an API-level property wrapper may be useful for wrappers that don't fit the current definition. The `@propertyWrapper` attribute could have an explicit `apiLevel` option that allows library authors to define whether the property wrapper has an external effect on the wrapped declaration:
+
+```swift
+@propertyWrapper(apiLevel)
+struct Asserted<Value> {
+  init(
+    wrappedValue: Value,
+    _ assertion: (Value) -> Bool,
+    file: StaticString = #file,
+    line: UInt = #line
+  ) { ... }
+}
+```
+
+### Library-defined diagnostic notes for understanding API wrapper restrictions
+
+API property wrappers on parameters have more restrictions than implementation detail property wrappers, because they have an external effect to the function. One very nice thing about property wrappers as custom attributes is that the documentation for that attribute is defined by the library author. Extend this principle, we could allow library authors to also define custom diagnostic notes to help explain restrictions that exist due to the external effects the wrapper imposes on the function.
+
+A custom diagnostic framework may also be useful for API-level property wrappers that are applied to properties, because such wrappers often change the semantics of the wrapped declaration in a way that fundamentally changes how the declaration can be used. For instance, a property wrapper may change the accessor mutability of the wrapped property, which may cause unexpected and unhelpful diagnostics when the programmer accesses the property incorrectly. The compiler cannot help explain _why_ there is a difference in mutability when the wrapper is attached, because the compiler does not understand how the property wrapper may change the behavior of the wrapped value.
+
+### API property wrappers in protocol requirements
+
+Protocol requirements that include property wrappers was [pitched](https://forums.swift.org/t/property-wrapper-requirements-in-protocols/33953) a while ago, but there was a lot of disagreement about whether property wrappers are implementation detail or API. With this distinction formalized, we could allow only API-level property wrappers in protocol requirements.
+
 ### Generalized property-wrapper initialization from a projection
```

Comment:
```suggestion
### Generalized projection initialization 
```
```suggestion
### Generalized projected-value initialization 
```

I think removing "property-wrapper" improves readability. We've it made clear throughout the proposal that this feature is extending property wrappers, so I think we can remove this redundancy.

Reply:
The phrase "property wrapper initialization from a projection" is deliberate because that's specifically referencing the syntactic sugar where you can assign to the `$` property and it will initialize the property wrapper. "projected-value initialization" sounds like regular initialization of projected value, which isn't the same thing.